internal/webhooks/machinedeployment_webhook.go

```diff
@@ -36,40 +36,48 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/feature"
 	"sigs.k8s.io/cluster-api/util/version"
 )
 
-func (m *MachineDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+// Webhook is the interface for all webhooks.
+type Webhook interface {
+	webhook.CustomDefaulter
+	webhook.CustomValidator
+	SetupWebhookWithManager(ctrl.Manager) error
+}
+
+func (webhook *machineDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
 	// This registers MachineDeployment as a validating webhook and
 	// machineDeploymentDefaulter as a defaulting webhook.
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(m).
-		WithDefaulter(MachineDeploymentDefaulter(mgr.GetScheme())).
+		For(&clusterv1.MachineDeployment{}).
+		WithDefaulter(webhook).
+		WithValidator(webhook).
 		Complete()
 }
 
 // +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1beta1-machinedeployment,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1beta1,name=validation.machinedeployment.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 // +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1beta1-machinedeployment,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1beta1,name=default.machinedeployment.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 
-var _ webhook.CustomDefaulter = &machineDeploymentDefaulter{}
-var _ webhook.Validator = &MachineDeployment{}
-
-// MachineDeploymentDefaulter creates a new CustomDefaulter for MachineDeployments.
-func MachineDeploymentDefaulter(scheme *runtime.Scheme) webhook.CustomDefaulter {
-	return &machineDeploymentDefaulter{
+// MachineDeploymentWebhook creates a new Webhook for MachineDeployments.
+func MachineDeploymentWebhook(scheme *runtime.Scheme) Webhook {
+	return &machineDeployment{
```

Comment:
WDYT about implementing this the same way as in: https://github.com/kubernetes-sigs/cluster-api/blob/ea0dc4da091b13bea47dbb6d41340c1f434bac5d/controlplane/kubeadm/internal/webhooks/scale.go#L35?

(i.e. we would drop the constructor, take the scheme from the manager in SetupWebhookWithManager and get rid of the interface, also basically this aligns it to the other webhooks)

Reply:
I think I implemented the previous version and for some reason I wanted to keep SetupWebhookWithManager on the MachineDeployment object, but now it seems better to be consistent to me

Reply:
I guess the only risk of that, is that you then may not have a decoder present right? And in all test we would have to set up the decoder and pass through, with this at least we have that guarantee.

I can get a commit up to compare and discuss though

Reply:
Yup. Seems a similar risk to that the Client is not set in the Cluster & ClusterClass webhooks (similar behavior in all of our controllers)

I think I'm not against ensuring that everything is there via constructors, I would mostly prefer consistency across webhooks (& ideally controllers)

Reply:
Basically we currently assume that webhooks & controllers are always used ~ like this:
```go
	if err := (&controllers.MachineReconciler{
		Client:                    mgr.GetClient(),
...
	}).SetupWithManager(ctx, mgr, concurrency(machineConcurrency)); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "Machine")
		os.Exit(1)
	}
...
	if err := (&webhooks.Cluster{Client: mgr.GetClient()}).SetupWebhookWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create webhook", "webhook", "Cluster")
		os.Exit(1)
	}
```

Reply:
Maybe one option is to aim for consistency across webhooks for this PR and (if we want) do a follow-up to figure out if we want to make it safer across the board

Reply:
I think at least the current layering with internal enforces that folks have to call SetupWebhookWithManager to get an instance of the internal webhook